protocol/directory.go

```diff
@@ -0,0 +1,109 @@
+package protocol
+
+import (
+	"bytes"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	"github.com/coniks-sys/coniks-go/merkletree"
+)
+
+type ConiksUserPolicies struct {
+	Username               string
+	AllowUnsignedKeychange bool
+	AllowPublicLookup      bool
+}
+
+type ConiksDirectory struct {
+	pad      *merkletree.PAD
+	useTBs   bool
+	tbs      map[string]*merkletree.TemporaryBinding
+	policies merkletree.Policies
+}
+
+func InitDirectory(epDeadline merkletree.TimeStamp, vrfKey *vrf.PrivateKey,
+	signKey sign.PrivateKey, dirSize uint64,
+	useTBs bool, capacity uint64) *ConiksDirectory {
+	d := new(ConiksDirectory)
+	d.SetPolicies(epDeadline, vrfKey)
+	pad, err := merkletree.NewPAD(d.policies, signKey, dirSize)
+	if err != nil {
+		panic(err)
+	}
+	d.pad = pad
+	d.useTBs = useTBs
+	if useTBs {
+		d.tbs = make(map[string]*merkletree.TemporaryBinding, capacity)
+	}
+	return d
+}
+
+func (d *ConiksDirectory) Update() {
+	d.pad.Update(d.policies)
+	// clear issued temporary bindings
+	for key := range d.tbs {
+		delete(d.tbs, key)
+	}
+}
+
+func (d *ConiksDirectory) SetPolicies(epDeadline merkletree.TimeStamp, vrfKey *vrf.PrivateKey) {
+	d.policies = merkletree.NewPolicies(epDeadline, vrfKey)
+}
+
+func (d *ConiksDirectory) LatestSTR() *merkletree.SignedTreeRoot {
+	return d.pad.LatestSTR()
+}
+
+func (d *ConiksDirectory) Register(uname string, key []byte) (
```

Comment:
Hmm, this is an interesting idea. I still think the more natural way to think about a registration and lookup would be to pass in the individual values, instead of the request message, though. I think the directory should remain a higher-level abstraction to the PAD, and it should still be up to the server to handle the message parsing and pass in the parsed values.